File: app/src/main/java/com/awscherb/cardkeeper/ui/card_detail/CardDetailPresenter.java
package com.awscherb.cardkeeper.ui.card_detail;

import com.awscherb.cardkeeper.data.service.ScannedCodeService;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;


public class CardDetailPresenter implements CardDetailContract.Presenter {

    private CardDetailContract.View view;

    @Inject ScannedCodeService service;
    private Disposable loadSubscription;

    //================================================================================
    // Constructor
    //================================================================================

    @Inject
    public CardDetailPresenter() {


    }

    //================================================================================
    // Presenter methods
    //================================================================================


    @Override
    public void attachView(CardDetailContract.View view) {
        this.view = view;
    }

    @Override
    public void loadCard(int id) {
        unsubscribe(loadSubscription);

        loadSubscription = service.getScannedCode(id)
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(view::showCard,
                        Throwable::printStackTrace);
    }

    @Override
    public void onViewDestroyed() {
        unsubscribe(loadSubscription);
    }

    private void unsubscribe(Disposable s) {
        if (s != null) {
            s.dispose();
        }
    }

}

File: app/src/main/java/com/awscherb/cardkeeper/ui/card_detail/CardDetailPresenter.kt
package com.awscherb.cardkeeper.ui.card_detail

import com.awscherb.cardkeeper.data.model.ScannedCode
import com.awscherb.cardkeeper.data.service.ScannedCodeService
import com.awscherb.cardkeeper.ui.base.Presenter
import kotlinx.coroutines.CoroutineDispatcher
import javax.inject.Inject

class CardDetailPresenter @Inject constructor(
    dispatcher: CoroutineDispatcher,
    private val service: ScannedCodeService
) : Presenter<CardDetailContract.View>(dispatcher), CardDetailContract.Presenter {

    private lateinit var originalTitle: String
    private lateinit var card: ScannedCode

    //================================================================================
    // Presenter methods
    //================================================================================

    override fun loadCard(id: Int) {
        uiScope(view::onError) {
            val code = service.getScannedCode(id)
            card = code
            originalTitle = code.title
            view.showCard(code)
        }
    }

    override fun setTitle(title: String) {
        view.setSaveVisible(title != originalTitle)
        card.title = title
    }

    override fun saveCard() {
        uiScope(view::onError) {
            service.updateScannedCode(card)
            view.onCardSaved()
        }
    }
}
